torch/_dynamo/trace_rules.py

```diff
@@ -35,40 +44,121 @@
+    "torch._dynamo.external_utils.is_compiling": TorchInGraphFunctionVariable,
+    "torch.autograd.graph.disable_saved_tensors_hooks": TorchInGraphFunctionVariable,
+}
+
+
+auto_torch_name_rule_map = {
+    # Dynamo implemented context managers
     "torch._C.DisableTorchFunctionSubclass": TorchCtxManagerClassVariable,
     "torch.amp.autocast_mode.autocast": TorchCtxManagerClassVariable,
     "torch.autograd.grad_mode.enable_grad": TorchCtxManagerClassVariable,
     "torch.autograd.grad_mode.inference_mode": TorchCtxManagerClassVariable,
     "torch.autograd.grad_mode.no_grad": TorchCtxManagerClassVariable,
     "torch.autograd.grad_mode.set_grad_enabled": TorchCtxManagerClassVariable,
-    "torch.autograd.profiler.profile": TorchCtxManagerClassVariable,
-    "torch.autograd.profiler.record_function": TorchCtxManagerClassVariable,
     "torch.cpu.amp.autocast_mode.autocast": TorchCtxManagerClassVariable,
     "torch.cuda.amp.autocast_mode.autocast": TorchCtxManagerClassVariable,
-    "torch.profiler.profiler.profile": TorchCtxManagerClassVariable,
 }
 
 
+torch_name_rule_map = {**manual_torch_name_rule_map, **auto_torch_name_rule_map}
+
+
 @functools.lru_cache(None)
 def get_torch_obj_rule_map():
     d = dict()
     for k, v in torch_name_rule_map.items():
-        obj = load_object(k)
-        assert obj not in d
-        d[obj] = v
+        try:
+            obj = load_object(k)
+            d[obj] = v
+        except (AttributeError, ModuleNotFoundError):
+            pass
     return d
 
 
+def _load_obj_from_str(fully_qualified_name):
+    module, obj_name = fully_qualified_name.rsplit(".", maxsplit=1)
+    return getattr(importlib.import_module(module), obj_name)
+
+
 def load_object(name):
-    mod_name, obj_name = name.rsplit(".", 1)
-    mod = importlib.import_module(mod_name)
-    obj = getattr(mod, obj_name)
-    return obj
+    x = name.split("#")
+    if len(x) == 2:
+        obj = _load_obj_from_str(x[0])
+        val = getattr(obj, x[1])
+    else:
+        assert len(x) == 1
+        val = _load_obj_from_str(x[0])
+    if hasattr(val, "__wrapped__") and val is not torch.ops:
+        val = val.__wrapped__
+    return val
+
+
+@functools.lru_cache(None)
+def get_tensor_method():
+    s = set()
+    for name in dir(torch.Tensor):
+        method = getattr(torch.Tensor, name)
+        if isinstance(
+            method, (types.MethodDescriptorType, types.WrapperDescriptorType)
+        ):
+            s.add(method)
+    return s
+
+
+def is_in_graph_function(obj):
+    if obj in get_tensor_method() or isinstance(
+        obj,
+        (torch._ops.OpOverloadPacket, torch._ops.OpOverload),
+    ):
+        return True
+    if isinstance(
+        obj,
+        (
+            types.FunctionType,
+            types.MethodType,
+            types.BuiltinFunctionType,
+            types.MethodDescriptorType,
+            types.WrapperDescriptorType,
+        ),
+    ):
+        return is_allowed(obj)
```

Comment:
Why can't we just unconditionally called `is_allowed`, no matter what kind of obj it is? If it's not one of these types, I'd expect it to just not be in `is_allowed`, no?

Reply:
No, we can't, because:
* We have to return ```TorchInGraphVariable``` for allowed functions, so we have to guard on these types.
* For other allowed, we return ```TorchVariable``` in this PR.

This function would be removed in the next PR when we hard coding the list of in graph functions, and just check the name_rule_map.

Reply:
No, I don't understand this explanation. There is no guarding here in the Dynamo sense, what do you mean by "guard on these types"

Reply:
Non function type, e.g., class, python module can be in ```is_allowed```:
```
>>> torch._dynamo.allowed_functions.is_allowed(torch.nn.modules)
True
>>> torch._dynamo.allowed_functions.is_allowed(torch.nn.Linear)
True
>>> 
```
This is why I have the condition, and why we need this huge refactor.

Reply:
Seems good to be written down somewhere